onnxscript/ir/_core.py

```diff
@@ -1027,11 +1072,67 @@ def opset_imports(self) -> dict[str, int]:
 
     @property
     def nodes(self) -> Sequence[Node]:
-        return self._nodes
+        return tuple(self._nodes)
 
-    def topologically_sorted_nodes(self) -> Sequence[Node]:
+    def __getitem__(self, index: int) -> Node:
+        return self._nodes[index]
+
+    def __len__(self) -> int:
+        return len(self._nodes)
+
+    def __iter__(self) -> Iterator[Node]:
+        return iter(self._nodes)
+
+    def __reversed__(self) -> Iterator[Node]:
+        return reversed(self._nodes)
+
+    def _remove_graph_reference_from_node(self, node: Node) -> None:
+        """A private function to remove the graph reference from a node, used when mutating the graph by node_list."""
+        # pylint: disable=protected-access
+        node._graph = None
+        # pylint: enable=protected-access
+
+    def _add_graph_reference_to_node(self, node: Node) -> None:
+        """A private function to add the graph reference to a node, used when mutating the graph by node_list."""
+        # pylint: disable=protected-access
+        if node._graph is not None:
+            node._graph.remove(node)
+        node._graph = self
+        # pylint: enable=protected-access
+
+    # Mutation methods
+    def append(self, node: Node) -> None:
+        """Append a node to the graph in O(1) time."""
+        self._nodes.append(node, property_modifier=self._add_graph_reference_to_node)
+
+    def extend(self, nodes: Iterable[Node]) -> None:
+        """Extend the graph with the given nodes in O(#new_nodes) time."""
+        self._nodes.extend(nodes, property_modifier=self._add_graph_reference_to_node)
+
+    def remove(self, node: Node) -> None:
+        """Remove a node from the graph in O(1) time."""
+        if node.graph is not self:
+            raise ValueError(f"The node {node} does not belong to this graph.")
+        self._nodes.remove(node, property_modifier=self._remove_graph_reference_from_node)
+
+    def insert_after(self, node: Node, new_nodes: Iterable[Node]) -> None:
+        """Insert new nodes after the given node in O(#new_nodes) time."""
+        self._nodes.insert_after(
+            node, new_nodes, property_modifier=self._add_graph_reference_to_node
+        )
+
+    def insert_before(self, node: Node, new_nodes: Iterable[Node]) -> None:
```

Comment:
How to insert only 1 node? `graph.insert_before(n, [new_node])`?

Reply:
That's correct. Any thoughts?

Reply:
have you considered supporting single node as argument as well and impl with isinstance check?

Reply:
> have you considered supporting single node as argument as well and impl with isinstance check?

We could but that could be added complexity. Do you have concrete use cases?

Reply:
> I see torch using
> 
> ```python
> with graph.inserting_after(node):
>     ...
> ```
> 
> Cause they have the need of creating nodes on the fly. I guess we don't have this need? Other than that, I think the imple here is simpler.

Pytorch needs the context manager because its graph is responsible for creating nodes. We don't need the context manager because our nodes are initialized independent of the graph, and can be inserted anywhere in any graph.

Reply:
Added an isinstance check to support single node